modules/bullet/shape_bullet.cpp

```diff
@@ -46,27 +46,56 @@
 	@author AndreaCatania
 */
 
-ShapeBullet::ShapeBullet() {}
+ShapeBullet::ShapeBullet() {
+}
 
-ShapeBullet::~ShapeBullet() {}
+ShapeBullet::~ShapeBullet() {
+	bulletdelete(default_shape);
+	default_shape = nullptr;
+}
 
 btCollisionShape *ShapeBullet::create_bt_shape(const Vector3 &p_implicit_scale, real_t p_extra_edge) {
 	btVector3 s;
 	G_TO_B(p_implicit_scale, s);
 	return create_bt_shape(s, p_extra_edge);
 }
 
+btCollisionShape *ShapeBullet::create_bt_shape(const btVector3 &p_implicit_scale, real_t p_extra_edge) {
+	if (p_extra_edge == 0.0 && (p_implicit_scale - btVector3(1, 1, 1)).length2() <= CMP_EPSILON) {
+		return default_shape;
+	}
+
+	return internal_create_bt_shape(p_implicit_scale, p_extra_edge);
+}
+
+void ShapeBullet::destroy_bt_shape(btCollisionShape *p_shape) const {
+	if (p_shape != default_shape && p_shape != old_default_shape) {
+		bulletdelete(p_shape);
+	}
+}
+
 btCollisionShape *ShapeBullet::prepare(btCollisionShape *p_btShape) const {
 	p_btShape->setUserPointer(const_cast<ShapeBullet *>(this));
 	p_btShape->setMargin(margin);
 	return p_btShape;
 }
 
 void ShapeBullet::notifyShapeChanged() {
+	// Store the old shape ptr so to not lose the reference pointer.
+	old_default_shape = default_shape;
+	// Create the new default shape with the new data.
+	default_shape = internal_create_bt_shape(btVector3(1, 1, 1));
+
 	for (Map<ShapeOwnerBullet *, int>::Element *E = owners.front(); E; E = E->next()) {
 		ShapeOwnerBullet *owner = static_cast<ShapeOwnerBullet *>(E->key());
 		owner->shape_changed(owner->find_shape(this));
 	}
+
+	if (old_default_shape) {
+		// At this point now one has the old default shape; just delete it.
+		bulletdelete(old_default_shape);
+		old_default_shape = nullptr;
+	}
 }
```

Comment:
It appears as if `old_default_shape` is only used (and cleared) here; so I would just make it a local variable to this function.

Reply:
The variable `old_default_shape` is used to avoid destroy the old default shape by the callbacks triggered by `owner->shape_changed(owner->find_shape(this));`